tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1178,186 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
+}
+
+/* Checks that a mutually-supported cipher suite can be used in this connection */
+static bool s2n_cipher_suite_match_is_valid(struct s2n_connection* conn, struct s2n_cipher_suite* potential_match) {
+    /* Never use TLS1.3 ciphers on a pre-TLS1.3 connection, and vice versa */
+    if ((conn->actual_protocol_version >= S2N_TLS13) != (potential_match->minimum_required_tls_version >= S2N_TLS13)) {
+        return false;
+    }
+
+    /* Skip the suite if we don't have an available implementation */
+    if (!potential_match->available) {
+        return false;
+    }
+
+    /* Make sure the cipher is valid for available certs */
+    if (s2n_is_cipher_suite_valid_for_auth(conn, potential_match) != S2N_SUCCESS) {
+        return false;
+    }
+
+    /* TLS 1.3 does not include key exchange in cipher suites */
+    if (potential_match->minimum_required_tls_version < S2N_TLS13) {
+        /* If the kex is not supported continue to the next candidate */
+        bool kex_supported = false;
+        if (s2n_result_is_error(s2n_kex_supported(potential_match, conn, &kex_supported))) {
+            return false;
+        }
+
+        if (!kex_supported) {
+            return false;
+        }
+
+        /* If the kex is not configured correctly continue to the next candidate */
+        if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+            return false;
+        }
+    }
+
+    /**
+    *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+    *# The server MUST ensure that it selects a compatible PSK
+    *# (if any) and cipher suite.
+    **/
+    if (conn->psk_params.chosen_psk != NULL) {
+        if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * Used by the server to select a mutually-supported cipher suite.
+ * This function iterates over the server's cipher suite preference list, choosing the first 
+ * suite that is also supported by the client. If an in-group deliminator is encountered, for 
+ * each in-group cipher suite, we iterate through the client's preferences and save the 
+ * index of the client's cipher suite, throwing out the previous index if it is 
+ * less-preferred by the client than the current index. Once the end-group deliminator is 
+ * encountered, the server either accepts the idenfitied cipher suite or if no cipher suite 
+ * has been found, continues iterating through its list. This function checks that the potential cipher
+ * suite match is valid and can be used for the connection.
+ */
+static S2N_RESULT s2n_get_mutually_supported_cipher(struct s2n_connection* conn,
+                                                    const uint8_t* wire,
+                                                    uint32_t count,
+                                                    uint32_t cipher_suite_len,
+                                                    struct s2n_cipher_suite** negotiated_cipher) {
```

Comment:
What does 'just wrap' mean?

Reply:
Hard to do in a github comment because github comments don't have reasonable text editor rules about line wraps, but I tried :)
```suggestion
static S2N_RESULT s2n_get_mutually_supported_cipher(struct s2n_connection* conn, const uint8_t* wire, uint32_t count, 
        uint32_t cipher_suite_len, struct s2n_cipher_suite** negotiated_cipher) {
```